src/handlers.rs

```diff
@@ -143,22 +144,46 @@ pub fn handle_request(
 
         RequestType::GETAI => {
             let hostname = CStr::from_bytes_with_nul(request.key)?.to_str()?;
-            let resp = dns_lookup::getaddrinfo(Some(hostname), None, None);
-
+            // Boths hints are necessary to mimick the glibc behaviour.
+            let hints = AddrInfoHints {
+                // The canonical name will be filled in the first
+                // addrinfo struct returned by getaddrinfo.
+                flags: AI_CANONNAME,
+                // Deduplicate the resulting addrs.
+                socktype: SOCK_STREAM,
+                address: 0,
+                protocol: 0,
+            };
+            let resp = dns_lookup::getaddrinfo(Some(hostname), None, Some(hints));
             let ai_resp_empty = AiResponse {
                 canon_name: hostname.to_string(),
                 addrs: vec![],
             };
-
             let ai_resp: AiResponse = match resp {
                 Ok(ai_resp_iter) => {
-                    let addrs: HashSet<IpAddr> = ai_resp_iter
+                    let name_and_addrs: Vec<(Option<String>, IpAddr)> = ai_resp_iter
```

Comment:
Not too proud of this part. Can you think of a more rust-idiomatic way to retrieve the addrs and the canonical name in one go?

As a reminder, the canonical name is stored in the first Airesponse struct.

Reply:
I'd write this block like this:


```rust
            let ai_resp: AiResponse = match resp {
                Ok(ai_resp_iter) => {
                    // filter out bad responses.
                    let mut ai_resp_iter = ai_resp_iter.filter_map(|e| e.ok()).peekable();
                    // According to man 3 getaddrinfo, the resulting
                    // canonical name should be stored in the first
                    // addrinfo struct.
                    // Re-using the request hostname if we don't get a
                    // canonical name.
                    let canon_name = ai_resp_iter
                        .peek()
                        .and_then(|e| e.canonname.to_owned())
                        .unwrap_or(hostname.to_string());

                    let mut addrs = ai_resp_iter.map(|e| e.sockaddr.ip()).collect::<Vec<_>>();

                    addrs.dedup();

                    AiResponse { canon_name, addrs }
                }
                Err(_) => ai_resp_empty,
            };
```

Reply:
Nice! Peek is exactly what I was looking for!